### PR TITLE
Preserve linearization of universes in presence of side-effects.

### PR DESCRIPTION
**Pull request by mattam82**
See PR #688 for a discussion. Linearization of contexts was relying on a hidden assumption, this is a try to make it explicit.

### CODE REVIEW

File: kernel/term_typing.ml [hunk 1]
```diff
@@ -248,10 +247,11 @@ let infer_declaration ~trust env kn dcl =
       let tyj = infer_type env typ in
       let proofterm =
         Future.chain ~pure:true body (fun ((body,uctx),side_eff) ->
+          let uctx = Univ.ContextSet.to_context uctx in
```
Comment (SkySkimmer) [inline kernel/term_typing.ml]: Why do this only to do `ContextSet.of_context` afterwards?

**Comment by SkySkimmer** [comment]
Is it possible to use the universes from effects to bypass the universe binder check? ie something like
```
Lemma foo@{i} : A.
Proof. foo_tac. Qed.
```
such that the resulting `foo` takes more than 1 universe.

**Comment by maximedenes** [comment]
@mattam82 What is the current status of this PR?

**Comment by mattam82** [comment]
I need to rebase and test #688 before looking at this again

**Comment by mattam82** [comment]
Superseded by #6483

**Pull request closed without merging**